### PR TITLE
Fix error in Rust 2018 + no_core environment

### PR DESCRIPTION
**Pull request by taiki-e**
Minimized reproduction: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=5b9f6c3026ec9d856699fa6dbd4361f0

This is a fix for the error that occurred in #58702.

r? @Centril

### CODE REVIEW

**Comment by Centril** [comment]
@petrochenkov Is this possibly due to hard-coding paths in HIR lowering? (can we stop doing that and instead introduce `lang_items` which are at least "principled hacks"?)

**Comment by eddyb** [comment]
@Centril You'd need to move lang item collection to before HIR lowering. Doable, but AFAIK it currently is still happening after HIR lowering.

File: src/libsyntax/std_inject.rs [hunk 1]
```diff
@@ -47,6 +47,10 @@ pub fn maybe_inject_crates_ref(
 
     // the first name in this list is the crate name of the crate with the prelude
     let names: &[&str] = if attr::contains_name(&krate.attrs, "no_core") {
+        // Use `crate::module` because `::module` is not allowed on the Rust 2018 edition.
+        if rust_2018 {
+            INJECTED_CRATE_NAME.with(|opt_name| opt_name.set(Some("crate")));
```
Comment (eddyb) [inline src/libsyntax/std_inject.rs]: Isn't the convention that `INJECTED_CRATE_NAME` of `None` corresponds to `#![no_core]`?
That means you can handle that case where `INJECTED_CRATE_NAME` is used, without changing this file.

**Comment by Centril** [comment]
> @Centril You'd need to move lang item collection to before HIR lowering. Doable, but AFAIK it currently is still happening after HIR lowering.

@eddyb What are the primary reasons why we haven't done so hitherto? perf? "no one has put in the work"? something else? Otherwise... It seems like something we should do then?

**Comment by eddyb** [comment]
> It seems like something we should do then?

Yes. I think @durka poked at this a year or two ago, but, yeah, I agree we should try again.

**Comment by petrochenkov** [comment]
Should built-in macros also search traits like `Default`/`Debug`/etc as lang items?
I don't see much difference with the `for` case, all those items are not tied to the language so much as "real" lang items like e.g. `Drop`, they only need to be found somehow.

**Comment by petrochenkov** [comment]
Proc macros face a very similar problem when they need to find items from a certain crate, and we also have `extern crate self as name` when we are *inside* that certain crate.

Same approach can be used here, I believe - `extern crate self as core` + `::core::option`/`::core::iter`/etc.

**Comment by petrochenkov** [comment]
The issue with this lowering-time desugaring is that the paths it generates not only need to be resolved, but they also should be "presentable" and look differently depending on `no_std` mode.

`::std::option::Option` should actually say `std` rather than `core` where it would be found as a lang item, that's why we have all this dance with `injected_crate_name`.

In this sense what this PR does is the correct solution.
I could perhaps be done a bit cleaner though, e.g. with `crate_root` being a `&str` rather than `Option<&str>` and initialized with `std_inject::injected_crate_name().unwrap_or("crate")`.

**Comment by petrochenkov** [comment]
Nah, this is good as is.
`crate::foo` works on both editions unlike `::core::foo` and `::crate` doesn't work so `PathRoot` needs to be removed anyway, so it probably can't be made more uniform.

@bors r+

File: src/librustc_resolve/lib.rs [hunk 2]
```diff
@@ -1697,9 +1697,10 @@ impl<'a> hir::lowering::Resolver for Resolver<'a> {
         components: &[&str],
         is_value: bool
     ) -> hir::Path {
-        let segments = iter::once(keywords::PathRoot.ident())
+        let segments = crate_root.as_ref().map(|_| keywords::PathRoot.ident()).into_iter()
             .chain(
-                crate_root.into_iter()
+                crate_root.or(Some("crate"))
+                    .into_iter()
```
Comment (eddyb) [inline src/librustc_resolve/lib.rs]: Can't we move this into the callers of `resolve_str_path`? Or I guess we want `None` to mean "this crate"? If the latter, this seems fine to me!

Comment (eddyb) [reply src/librustc_resolve/lib.rs]: Although I'd very much prefer to see `keywords::Crate` rather than `"crate"` that gets interned later.

**Comment by Centril** [comment]
Beta nominating for bootstrap compiler due to https://github.com/rust-lang/rust/pull/58702#issuecomment-478210863.

**Comment by pnkfelix** [comment]
discussed at T-compiler meeting. Declining to beta-backport: The only client a backport would benefit would be `rustc` bootstraps (and not the stable channel), and we're already going to have nightly roll to beta next week anyway, so the risk/reward ratio is bad.

**Pull request merged**